fdb-extensions/src/main/java/com/apple/foundationdb/async/RankedSet.java

```diff
@@ -643,19 +819,21 @@ private static void checkKey(byte[] key) {
                         .thenApply(longs -> longs.stream().reduce(0L, Long::sum)));
     }
 
-    private CompletableFuture<byte[]> getPreviousKey(TransactionContext tc, int level, byte[] key) {
+    // Get the key before this one at the given level.
+    // If orEqual is given, then an exactly matching key is also considered. This is only used when the key is known
+    // to be a duplicate or an existing key and so should do whatever it did.
+    private CompletableFuture<byte[]> getPreviousKey(TransactionContext tc, int level, byte[] key, boolean orEqual) {
         byte[] k = subspace.pack(Tuple.from(level, key));
         CompletableFuture<byte[]> kf = tc.run(tr ->
-                tr.snapshot()
-                        .getRange(KeySelector.lastLessThan(k), KeySelector.firstGreaterOrEqual(k), 1)
-                        .asList()
-                        .thenApply(kvs -> {
-                            byte[] prevk = kvs.get(0).getKey();
-                            // If another key were inserted after between this and the target key,
-                            // it wouldn't be the one we should increment any more.
-                            // But do not conflict when key itself is incremented.
-                            byte[] exclusiveBegin = ByteArrayUtil.join(prevk, ZERO_ARRAY);
-                            tr.addReadConflictRange(exclusiveBegin, k);
+                tr.snapshot().getKey(orEqual ? KeySelector.lastLessOrEqual(k) : KeySelector.lastLessThan(k))
```

Comment:
I don't _think_ there was any reason to do this with a `getRange` before. I'm also not sure the `snapshot` is needed any more.

Reply:
The previous code,
```
.getRange(KeySelector.lastLessThan(k), KeySelector.firstGreaterOrEqual(k), 1)
```
didn't really have that protection, since it used two key selectors from the same input.

Would it be better then (that is, protecting against some screwup with the `('',0)` entry) to do a reverse `getRange` with the left the level limit and the right the target?

Reply:
Yeah, I think a range read like that would make sense, and it would protect us from exactly that kind of screw up.

Reply:
I switched to not using `KeySelector`s and doing a size 1 reverse range scan.